<!DOCTYPE html>
<html lang="en" data-accent-color="violet" data-content_root="../../">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0"><title>ginkgo_ai_client.queries - ginkgo_ai_client documentation</title><link rel="index" title="Index" href="../../genindex.html" /><link rel="search" title="Search" href="../../search.html" /><script>
    function setColorMode(t){let e=document.documentElement;e.setAttribute("data-color-mode",t);let a=window.matchMedia&&window.matchMedia("(prefers-color-scheme: dark)").matches,s=t;"auto"===t&&(s=a?"dark":"light"),"light"===s?(e.classList.remove("dark"),e.classList.add("light")):(e.classList.remove("light"),e.classList.add("dark"))}
    setColorMode(localStorage._theme||"auto");
  </script><link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=397bb51e" />
    <link rel="stylesheet" type="text/css" href="../../_static/shibuya.css?v=83eaa723" />
    <link media="print" rel="stylesheet" type="text/css" href="../../_static/print.css?v=20ff2c19" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css?v=b491029c" />
    <link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap" rel="stylesheet">
<style>
:root {
  --sy-f-text: "Inter", var(--sy-f-sys), var(--sy-f-cjk), sans-serif;
  --sy-f-heading: "Inter", var(--sy-f-sys), var(--sy-f-cjk), sans-serif;
}
</style>
    <meta property="og:type" content="website"/><meta property="og:title" content="ginkgo_ai_client.queries"/>
<meta name="twitter:card" content="summary"/>
  </head>
<body><div class="sy-head">
  <div class="sy-head-blur"></div>
  <div class="sy-head-inner sy-container mx-auto">
    <a class="sy-head-brand" href="../../index.html">
      
      
      <strong>ginkgo_ai_client</strong>
    </a>
    <div class="sy-head-nav" id="HeadNav">
      <nav class="sy-head-links"></nav>
      <div class="sy-head-extra flex items-center print:hidden"><form class="searchbox flex items-center" action="../../search.html" method="get">
  <input type="text" name="q" placeholder="Search" />
  <kbd>/</kbd>
</form><div class="sy-head-socials"></div></div>
    </div>
    <div class="sy-head-actions flex items-center shrink-0 print:hidden"><button class="js-theme theme-switch flex items-center"
data-aria-auto="Switch to light color mode"
data-aria-light="Switch to dark color mode"
data-aria-dark="Switch to auto color mode">
<i class="i-lucide theme-icon"></i>
</button><button class="md:hidden flex items-center js-menu" aria-label="Menu" type="button" aria-controls="HeadNav" aria-expanded="false">
        <div class="hamburger">
          <span class="hamburger_1"></span>
          <span class="hamburger_2 -translate-x-2"></span>
          <span class="hamburger_3 -translate-x-1"></span>
        </div>
      </button>
    </div>
  </div>
</div>
<div class="sy-page sy-container flex mx-auto">
  <aside id="lside" class="sy-lside md:w-72 md:shrink-0 print:hidden">
    <div class="sy-lside-inner md:sticky">
      <div class="sy-scrollbar p-6">
        <div class="globaltoc" data-expand-depth="0"><p class="caption" role="heading" aria-level="3"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../ref.html">API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Examples</a></li>
</ul>

        </div>
      </div>
    </div>
  </aside>
  <div class="lside-overlay js-menu" role="button" aria-label="Close left sidebar" aria-controls="lside" aria-expanded="false"></div>
  <aside id="rside" class="sy-rside pb-3 w-64 shrink-0 order-last">
    <button class="rside-close js-menu xl:hidden" aria-label="Close Table of Contents" type="button" aria-controls="rside" aria-expanded="false">
      <i class="i-lucide close"></i>
    </button>
    <div class="sy-scrollbar sy-rside-inner px-6 xl:top-16 xl:sticky xl:pl-0 pt-6 pb-4"><div id="ethical-ad-placement" data-ea-publisher="readthedocs"></div></div>
  </aside>
  <div class="rside-overlay js-menu" role="button" aria-label="Close Table of Contents" aria-controls="rside" aria-expanded="false"></div>
  <main class="sy-main w-full max-sm:max-w-full print:pt-6"><div class="sy-breadcrumbs" role="navigation">
  <div class="sy-breadcrumbs-inner flex items-center">
    <div class="md:hidden mr-3">
      <button class="js-menu" aria-label="Menu" type="button" aria-controls="lside" aria-expanded="false">
        <i class="i-lucide menu"></i>
      </button>
    </div>
    <ol class="flex-1" itemscope itemtype="https://schema.org/BreadcrumbList"><li itemprop="itemListElement" itemscope itemtype="https://schema.org/ListItem">
        <a itemprop="item" href="../../index.html"><span itemprop="name">ginkgo_ai_client</span></a>
        <span>/</span>
        <meta itemprop="position" content="1" />
      </li><li itemprop="itemListElement" itemscope itemtype="https://schema.org/ListItem">
        <a itemprop="item" href="../index.html"><span itemprop="name">Module code</span></a>
        <span>/</span>
        <meta itemprop="position" content="2" />
      </li><li itemprop="itemListElement" itemscope itemtype="https://schema.org/ListItem">
        <strong itemprop="name">ginkgo_ai_client.queries</strong>
        <meta itemprop="position" content="3" />
      </li></ol>
    <div class="xl:hidden ml-1">
      <button class="js-menu" aria-label="Show table of contents" type="button" aria-controls="rside"
        aria-expanded="false">
        <i class="i-lucide outdent"></i>
      </button>
    </div>
  </div>
</div><div class="flex flex-col break-words justify-between">
      <div class="min-w-0 max-w-6xl px-6 pb-6 pt-8 xl:px-12">
        <article class="yue" role="main">
          <h1>Source code for ginkgo_ai_client.queries</h1><div class="highlight"><pre>
<span></span><span data-line="1"><span class="sd">&quot;&quot;&quot;Classes to define queries to the Ginkgo AI API.&quot;&quot;&quot;</span>
</span><span data-line="2">
</span><span data-line="3"><span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">List</span>
</span><span data-line="4"><span class="kn">from</span> <span class="nn">abc</span> <span class="kn">import</span> <span class="n">ABC</span><span class="p">,</span> <span class="n">abstractmethod</span>
</span><span data-line="5"><span class="kn">import</span> <span class="nn">json</span>
</span><span data-line="6">
</span><span data-line="7"><span class="kn">import</span> <span class="nn">pydantic</span>
</span><span data-line="8">
</span><span data-line="9"><span class="kn">from</span> <span class="nn">ginkgo_ai_client.utils</span> <span class="kn">import</span> <span class="n">fasta_sequence_iterator</span><span class="p">,</span> <span class="n">IteratorWithLength</span>
</span><span data-line="10">
</span><span data-line="11">
</span><span data-line="12"><span class="k">class</span> <span class="nc">QueryBase</span><span class="p">(</span><span class="n">pydantic</span><span class="o">.</span><span class="n">BaseModel</span><span class="p">,</span> <span class="n">ABC</span><span class="p">):</span>
</span><span data-line="13"><span class="w">    </span><span class="sd">&quot;&quot;&quot;Base class for all queries. It&#39;s functions are:</span>
</span><span data-line="14"><span class="sd">    - Specify the mandatory class methods `to_request_params` and `parse_response`</span>
</span><span data-line="15"><span class="sd">    - Provide a better error message when a user forgets to use named arguments only.</span>
</span><span data-line="16"><span class="sd">      Without that tweak, the default error message from pydantic is very technical</span>
</span><span data-line="17"><span class="sd">      and confusing to new users.</span>
</span><span data-line="18"><span class="sd">    &quot;&quot;&quot;</span>
</span><span data-line="19">
</span><span data-line="20">    <span class="k">def</span> <span class="fm">__new__</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
</span><span data-line="21">        <span class="k">if</span> <span class="n">args</span><span class="p">:</span>
</span><span data-line="22">            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
</span><span data-line="23">                <span class="sa">f</span><span class="s2">&quot;Invalid initialization: </span><span class="si">{</span><span class="bp">cls</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2"> does not accept unnamed &quot;</span>
</span><span data-line="24">                <span class="sa">f</span><span class="s2">&quot;arguments. Please name all inputs, for instance &quot;</span>
</span><span data-line="25">                <span class="sa">f</span><span class="s2">&quot;`</span><span class="si">{</span><span class="bp">cls</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">(field_name=value, other_field=value, ...)`.&quot;</span>
</span><span data-line="26">            <span class="p">)</span>
</span><span data-line="27">        <span class="k">return</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span>
</span><span data-line="28">
</span><span data-line="29">    <span class="nd">@abstractmethod</span>
</span><span data-line="30">    <span class="k">def</span> <span class="nf">to_request_params</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
</span><span data-line="31">        <span class="k">pass</span>
</span><span data-line="32">
</span><span data-line="33">    <span class="nd">@abstractmethod</span>
</span><span data-line="34">    <span class="k">def</span> <span class="nf">parse_response</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">results</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Any</span><span class="p">:</span>
</span><span data-line="35">        <span class="k">pass</span>
</span><span data-line="36">
</span><span data-line="37">
</span><span data-line="38"><span class="k">class</span> <span class="nc">ResponseBase</span><span class="p">(</span><span class="n">pydantic</span><span class="o">.</span><span class="n">BaseModel</span><span class="p">):</span>
</span><span data-line="39">
</span><span data-line="40">    <span class="k">def</span> <span class="nf">write_to_jsonl</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
</span><span data-line="41">        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
</span><span data-line="42">            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_dump_json</span><span class="p">()</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span data-line="43">
</span><span data-line="44">
</span><span data-line="45"><span class="n">_maskedlm_models_properties</span> <span class="o">=</span> <span class="p">{</span>
</span><span data-line="46">    <span class="s2">&quot;ginkgo-aa0-650M&quot;</span><span class="p">:</span> <span class="s2">&quot;protein&quot;</span><span class="p">,</span>
</span><span data-line="47">    <span class="s2">&quot;esm2-650M&quot;</span><span class="p">:</span> <span class="s2">&quot;protein&quot;</span><span class="p">,</span>
</span><span data-line="48">    <span class="s2">&quot;esm2-3B&quot;</span><span class="p">:</span> <span class="s2">&quot;protein&quot;</span><span class="p">,</span>
</span><span data-line="49">    <span class="s2">&quot;ginkgo-maskedlm-3utr-v1&quot;</span><span class="p">:</span> <span class="s2">&quot;dna&quot;</span><span class="p">,</span>
</span><span data-line="50">    <span class="s2">&quot;lcdna&quot;</span><span class="p">:</span> <span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span>
</span><span data-line="51">    <span class="s2">&quot;abdiffusion&quot;</span><span class="p">:</span> <span class="s2">&quot;protein&quot;</span><span class="p">,</span>
</span><span data-line="52"><span class="p">}</span>
</span><span data-line="53">
</span><span data-line="54"><span class="n">_maskedlm_models_properties_str</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
</span><span data-line="55">    <span class="sa">f</span><span class="s2">&quot;- </span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">sequence_type</span><span class="si">}</span><span class="s2">&quot;</span>
</span><span data-line="56">    <span class="k">for</span> <span class="n">model</span><span class="p">,</span> <span class="n">sequence_type</span> <span class="ow">in</span> <span class="n">_maskedlm_models_properties</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
</span><span data-line="57"><span class="p">)</span>
</span><span data-line="58">
</span><span data-line="59">
</span><span data-line="60"><span class="k">def</span> <span class="nf">_validate_model_and_sequence</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">allow_masks</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span data-line="61"><span class="w">    </span><span class="sd">&quot;&quot;&quot;Raise an error if the model is unknown or the sequence isn&#39;t compatible.&quot;&quot;&quot;</span>
</span><span data-line="62">    <span class="n">valid_models</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">_maskedlm_models_properties</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</span><span data-line="63">    <span class="k">if</span> <span class="n">model</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_models</span><span class="p">:</span>
</span><span data-line="64">        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Model &#39;</span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s2">&#39; unknown. Sould be one of </span><span class="si">{</span><span class="n">valid_models</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span data-line="65">    <span class="n">sequence_type</span> <span class="o">=</span> <span class="n">_maskedlm_models_properties</span><span class="p">[</span><span class="n">model</span><span class="p">]</span>
</span><span data-line="66">    <span class="k">if</span> <span class="n">allow_masks</span><span class="p">:</span>
</span><span data-line="67">        <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&lt;mask&gt;&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
</span><span data-line="68">    <span class="k">if</span> <span class="n">sequence_type</span> <span class="o">==</span> <span class="s2">&quot;dna&quot;</span><span class="p">:</span>
</span><span data-line="69">        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">}):</span>
</span><span data-line="70">            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
</span><span data-line="71">                <span class="sa">f</span><span class="s2">&quot;Model </span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s2"> requires the sequence to only contain ATGC characters&quot;</span>
</span><span data-line="72">            <span class="p">)</span>
</span><span data-line="73">    <span class="k">elif</span> <span class="n">sequence_type</span> <span class="o">==</span> <span class="s2">&quot;nucleotide&quot;</span><span class="p">:</span>
</span><span data-line="74">        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">sequence</span><span class="o">.</span><span class="n">lower</span><span class="p">())</span><span class="o">.</span><span class="n">issubset</span><span class="p">({</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;s&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="s2">&quot;k&quot;</span><span class="p">,</span> <span class="s2">&quot;m&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span> <span class="s2">&quot;h&quot;</span><span class="p">,</span> <span class="s2">&quot;v&quot;</span><span class="p">,</span> <span class="s2">&quot;n&quot;</span><span class="p">}):</span>
</span><span data-line="75">            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
</span><span data-line="76">                <span class="sa">f</span><span class="s2">&quot;Model </span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s2"> requires the sequence to only contain valid IUPAC nucleotide characters&quot;</span>
</span><span data-line="77">            <span class="p">)</span>
</span><span data-line="78">    <span class="k">elif</span> <span class="n">sequence_type</span> <span class="o">==</span> <span class="s2">&quot;protein&quot;</span><span class="p">:</span>
</span><span data-line="79">        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="s2">&quot;ACDEFGHIKLMNPQRSTVWY&quot;</span><span class="p">)):</span>
</span><span data-line="80">            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Sequence must contain only protein characters&quot;</span><span class="p">)</span>
</span><span data-line="81">    <span class="k">else</span><span class="p">:</span>
</span><span data-line="82">        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid sequence type&quot;</span><span class="p">)</span>
</span><span data-line="83">
</span><span data-line="84">
<div class="viewcode-block" id="EmbeddingResponse">
<a class="viewcode-back" href="../../ref.html#ginkgo_ai_client.queries.EmbeddingResponse">[docs]</a>
</span><span data-line="85"><span class="k">class</span> <span class="nc">EmbeddingResponse</span><span class="p">(</span><span class="n">ResponseBase</span><span class="p">):</span>
</span><span data-line="86"><span class="w">    </span><span class="sd">&quot;&quot;&quot;A response to a MeanEmbeddingQuery, with attributes `embedding` (the mean</span>
</span><span data-line="87"><span class="sd">    embedding of the model&#39;s last encoder layer) and `query_name` (the original</span>
</span><span data-line="88"><span class="sd">    query&#39;s name).</span>
</span><span data-line="89"><span class="sd">    &quot;&quot;&quot;</span>
</span><span data-line="90">
</span><span data-line="91">    <span class="n">embedding</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span>
</span><span data-line="92">    <span class="n">query_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span></div>

</span><span data-line="93">
</span><span data-line="94">
<div class="viewcode-block" id="MeanEmbeddingQuery">
<a class="viewcode-back" href="../../ref.html#ginkgo_ai_client.queries.MeanEmbeddingQuery">[docs]</a>
</span><span data-line="95"><span class="k">class</span> <span class="nc">MeanEmbeddingQuery</span><span class="p">(</span><span class="n">QueryBase</span><span class="p">):</span>
</span><span data-line="96"><span class="w">    </span><span class="sd">&quot;&quot;&quot;A query to infer mean embeddings from a DNA or protein sequence.</span>
</span><span data-line="97">
</span><span data-line="98"><span class="sd">    Parameters</span>
</span><span data-line="99"><span class="sd">    ----------</span>
</span><span data-line="100"><span class="sd">    sequence: str</span>
</span><span data-line="101"><span class="sd">        The sequence to unmask. The sequence should be of the form &quot;MLPP&lt;mask&gt;PPLM&quot; with</span>
</span><span data-line="102"><span class="sd">        as many masks as desired.</span>
</span><span data-line="103"><span class="sd">    model: str</span>
</span><span data-line="104"><span class="sd">        The model to use for the inference.</span>
</span><span data-line="105"><span class="sd">    query_name: Optional[str] = None</span>
</span><span data-line="106"><span class="sd">        The name of the query. It will appear in the API response and can be used to</span>
</span><span data-line="107"><span class="sd">        handle exceptions.</span>
</span><span data-line="108">
</span><span data-line="109"><span class="sd">    Returns</span>
</span><span data-line="110"><span class="sd">    -------</span>
</span><span data-line="111"><span class="sd">    EmbeddingResponse</span>
</span><span data-line="112"><span class="sd">        ``client.send_request(query)`` returns an ``EmbeddingResponse`` with attributes</span>
</span><span data-line="113"><span class="sd">        ``embedding`` (the mean embedding of the model&#39;s last encoder layer) and</span>
</span><span data-line="114"><span class="sd">        ``query_name`` (the original query&#39;s name).</span>
</span><span data-line="115">
</span><span data-line="116"><span class="sd">    Examples</span>
</span><span data-line="117"><span class="sd">    --------</span>
</span><span data-line="118"><span class="sd">    &gt;&gt;&gt; query = MeanEmbeddingQuery(&quot;MLPP&lt;mask&gt;PPLM&quot;, model=&quot;ginkgo-aa0-650M&quot;)</span>
</span><span data-line="119"><span class="sd">    &gt;&gt;&gt; client.send_request(query)</span>
</span><span data-line="120"><span class="sd">    EmbeddingResponse(embedding=[1.05, 0.002, ...])</span>
</span><span data-line="121"><span class="sd">    &quot;&quot;&quot;</span>
</span><span data-line="122">
</span><span data-line="123">    <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span>
</span><span data-line="124">    <span class="n">model</span><span class="p">:</span> <span class="nb">str</span>
</span><span data-line="125">    <span class="n">query_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span data-line="126">
</span><span data-line="127">    <span class="k">def</span> <span class="nf">to_request_params</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
</span><span data-line="128">        <span class="k">return</span> <span class="p">{</span>
</span><span data-line="129">            <span class="s2">&quot;model&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span>
</span><span data-line="130">            <span class="s2">&quot;text&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span>
</span><span data-line="131">            <span class="s2">&quot;transforms&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;EMBEDDING&quot;</span><span class="p">}],</span>
</span><span data-line="132">        <span class="p">}</span>
</span><span data-line="133">
</span><span data-line="134">    <span class="k">def</span> <span class="nf">parse_response</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">results</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">EmbeddingResponse</span><span class="p">:</span>
</span><span data-line="135">        <span class="k">return</span> <span class="n">EmbeddingResponse</span><span class="p">(</span>
</span><span data-line="136">            <span class="n">embedding</span><span class="o">=</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;embedding&quot;</span><span class="p">],</span> <span class="n">query_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">query_name</span>
</span><span data-line="137">        <span class="p">)</span>
</span><span data-line="138">
</span><span data-line="139">    <span class="nd">@pydantic</span><span class="o">.</span><span class="n">model_validator</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s2">&quot;after&quot;</span><span class="p">)</span>
</span><span data-line="140">    <span class="k">def</span> <span class="nf">check_model_and_sequence_compatibility</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">query</span><span class="p">):</span>
</span><span data-line="141">        <span class="n">sequence</span><span class="p">,</span> <span class="n">model</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="n">query</span><span class="o">.</span><span class="n">model</span>
</span><span data-line="142">        <span class="n">_validate_model_and_sequence</span><span class="p">(</span><span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span> <span class="n">allow_masks</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</span><span data-line="143">        <span class="k">return</span> <span class="n">query</span>
</span><span data-line="144">
</span><span data-line="145">    <span class="nd">@classmethod</span>
</span><span data-line="146">    <span class="k">def</span> <span class="nf">iter_from_fasta</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">fasta_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
</span><span data-line="147"><span class="w">        </span><span class="sd">&quot;&quot;&quot;Return an iterator over the sequences in a fasta file. The iterator has</span>
</span><span data-line="148"><span class="sd">        a length attribute that gives the number of sequences in the fasta file.&quot;&quot;&quot;</span>
</span><span data-line="149">        <span class="n">fasta_iterator</span> <span class="o">=</span> <span class="n">fasta_sequence_iterator</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">)</span>
</span><span data-line="150">        <span class="n">query_iterator</span> <span class="o">=</span> <span class="p">(</span>
</span><span data-line="151">            <span class="bp">cls</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span> <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> <span class="n">query_name</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
</span><span data-line="152">            <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fasta_iterator</span>
</span><span data-line="153">        <span class="p">)</span>
</span><span data-line="154">        <span class="k">return</span> <span class="n">IteratorWithLength</span><span class="p">(</span><span class="n">query_iterator</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">fasta_iterator</span><span class="p">))</span>
</span><span data-line="155">
</span><span data-line="156">    <span class="nd">@classmethod</span>
</span><span data-line="157">    <span class="k">def</span> <span class="nf">list_from_fasta</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">fasta_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
</span><span data-line="158">        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">cls</span><span class="o">.</span><span class="n">iter_from_fasta</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="n">model</span><span class="p">))</span></div>

</span><span data-line="159">
</span><span data-line="160">
<div class="viewcode-block" id="SequenceResponse">
<a class="viewcode-back" href="../../ref.html#ginkgo_ai_client.queries.SequenceResponse">[docs]</a>
</span><span data-line="161"><span class="k">class</span> <span class="nc">SequenceResponse</span><span class="p">(</span><span class="n">ResponseBase</span><span class="p">):</span>
</span><span data-line="162"><span class="w">    </span><span class="sd">&quot;&quot;&quot;A response to a MaskedInferenceQuery, with attributes `sequence` (the predicted</span>
</span><span data-line="163"><span class="sd">    sequence) and `query_name` (the original query&#39;s name).</span>
</span><span data-line="164"><span class="sd">    &quot;&quot;&quot;</span>
</span><span data-line="165">
</span><span data-line="166">    <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span>
</span><span data-line="167">    <span class="n">query_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span></div>

</span><span data-line="168">
</span><span data-line="169">
<div class="viewcode-block" id="MaskedInferenceQuery">
<a class="viewcode-back" href="../../ref.html#ginkgo_ai_client.queries.MaskedInferenceQuery">[docs]</a>
</span><span data-line="170"><span class="k">class</span> <span class="nc">MaskedInferenceQuery</span><span class="p">(</span><span class="n">QueryBase</span><span class="p">):</span>
</span><span data-line="171"><span class="w">    </span><span class="sd">&quot;&quot;&quot;A query to infer masked tokens in a DNA or protein sequence.</span>
</span><span data-line="172">
</span><span data-line="173"><span class="sd">    Parameters</span>
</span><span data-line="174"><span class="sd">    ----------</span>
</span><span data-line="175"><span class="sd">    sequence: str</span>
</span><span data-line="176"><span class="sd">        The sequence to unmask. The sequence should be of the form &quot;MLPP&lt;mask&gt;PPLM&quot; with</span>
</span><span data-line="177"><span class="sd">        as many masks as desired.</span>
</span><span data-line="178"><span class="sd">    model: str</span>
</span><span data-line="179"><span class="sd">        The model to use for the inference (only &quot;ginkgo-aa0-650M&quot; is supported for now).</span>
</span><span data-line="180"><span class="sd">    query_name: Optional[str] = None</span>
</span><span data-line="181"><span class="sd">        The name of the query. It will appear in the API response and can be used to</span>
</span><span data-line="182"><span class="sd">        handle exceptions.</span>
</span><span data-line="183">
</span><span data-line="184"><span class="sd">    Returns</span>
</span><span data-line="185"><span class="sd">    --------</span>
</span><span data-line="186"><span class="sd">    SequenceResponse</span>
</span><span data-line="187"><span class="sd">        ``client.send_request(query)`` returns a ``SequenceResponse`` with attributes</span>
</span><span data-line="188"><span class="sd">        ``sequence` (the predicted sequence) and ``query_name`` (the original query&#39;s</span>
</span><span data-line="189"><span class="sd">        name).</span>
</span><span data-line="190">
</span><span data-line="191"><span class="sd">    &quot;&quot;&quot;</span>
</span><span data-line="192">
</span><span data-line="193">    <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span>
</span><span data-line="194">    <span class="n">model</span><span class="p">:</span> <span class="nb">str</span>
</span><span data-line="195">    <span class="n">query_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span data-line="196">
</span><span data-line="197">    <span class="k">def</span> <span class="nf">to_request_params</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
</span><span data-line="198">        <span class="k">return</span> <span class="p">{</span>
</span><span data-line="199">            <span class="s2">&quot;model&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span>
</span><span data-line="200">            <span class="s2">&quot;text&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span>
</span><span data-line="201">            <span class="s2">&quot;transforms&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;FILL_MASK&quot;</span><span class="p">}],</span>
</span><span data-line="202">        <span class="p">}</span>
</span><span data-line="203">
</span><span data-line="204">    <span class="k">def</span> <span class="nf">parse_response</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">response</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SequenceResponse</span><span class="p">:</span>
</span><span data-line="205"><span class="w">        </span><span class="sd">&quot;&quot;&quot;The response has a sequence and the original query&#39;s name&quot;&quot;&quot;</span>
</span><span data-line="206">        <span class="k">return</span> <span class="n">SequenceResponse</span><span class="p">(</span>
</span><span data-line="207">            <span class="n">sequence</span><span class="o">=</span><span class="n">response</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">],</span> <span class="n">query_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">query_name</span>
</span><span data-line="208">        <span class="p">)</span>
</span><span data-line="209">
</span><span data-line="210">    <span class="nd">@pydantic</span><span class="o">.</span><span class="n">model_validator</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s2">&quot;after&quot;</span><span class="p">)</span>
</span><span data-line="211">    <span class="k">def</span> <span class="nf">check_model_and_sequence_compatibility</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">query</span><span class="p">):</span>
</span><span data-line="212">        <span class="n">sequence</span><span class="p">,</span> <span class="n">model</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="n">query</span><span class="o">.</span><span class="n">model</span>
</span><span data-line="213">        <span class="n">_validate_model_and_sequence</span><span class="p">(</span><span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span> <span class="n">allow_masks</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span data-line="214">        <span class="k">return</span> <span class="n">query</span></div>

</span><span data-line="215">
</span><span data-line="216">
</span><span data-line="217"><span class="n">auto_doc_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
</span><span data-line="218"><span class="s2">    Supported inference models</span>
</span><span data-line="219"><span class="s2">    --------------------------</span>
</span><span data-line="220">
</span><span data-line="221"><span class="s2">    Here are the supported models, and the sequence type they support. Sequences must</span>
</span><span data-line="222"><span class="s2">    be upper-case and not contain any mask etc. for embeddings computation.</span>
</span><span data-line="223">
</span><span data-line="224"><span class="s2">    </span><span class="si">{</span><span class="n">_maskedlm_models_properties_str</span><span class="si">}</span>
</span><span data-line="225"><span class="s2">&quot;&quot;&quot;</span>
</span><span data-line="226">
</span><span data-line="227"><span class="k">for</span> <span class="bp">cls</span> <span class="ow">in</span> <span class="p">[</span><span class="n">MeanEmbeddingQuery</span><span class="p">,</span> <span class="n">MaskedInferenceQuery</span><span class="p">]:</span>
</span><span data-line="228">    <span class="bp">cls</span><span class="o">.</span><span class="vm">__doc__</span> <span class="o">+=</span> <span class="n">auto_doc_str</span><span class="p">[:</span><span class="mi">1</span><span class="p">]</span>
</span><span data-line="229">
</span><span data-line="230">
</span><span data-line="231"><span class="c1">### PROMOTER ACTIVITY QUERIES ###</span>
</span><span data-line="232">
</span><span data-line="233">
</span><span data-line="234"><span class="k">class</span> <span class="nc">PromoterActivityResponse</span><span class="p">(</span><span class="n">ResponseBase</span><span class="p">):</span>
</span><span data-line="235"><span class="w">    </span><span class="sd">&quot;&quot;&quot;A response to a PromoterActivityQuery, with attributes `activity` (the predicted</span>
</span><span data-line="236"><span class="sd">    activity) and `query_name` (the original query&#39;s name).</span>
</span><span data-line="237">
</span><span data-line="238"><span class="sd">    Attributes</span>
</span><span data-line="239"><span class="sd">    ----------</span>
</span><span data-line="240"><span class="sd">    activity_by_tissue: Dict[str, float]</span>
</span><span data-line="241"><span class="sd">        The activity of the promoter in each tissue.</span>
</span><span data-line="242"><span class="sd">    query_name: Optional[str] = None</span>
</span><span data-line="243"><span class="sd">        The name of the query. It will appear in the API response and can be used to</span>
</span><span data-line="244"><span class="sd">        handle exceptions.</span>
</span><span data-line="245"><span class="sd">    &quot;&quot;&quot;</span>
</span><span data-line="246">
</span><span data-line="247">    <span class="n">activity_by_tissue</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span>
</span><span data-line="248">    <span class="n">query_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span data-line="249">
</span><span data-line="250">
</span><span data-line="251"><span class="k">class</span> <span class="nc">PromoterActivityQuery</span><span class="p">(</span><span class="n">QueryBase</span><span class="p">):</span>
</span><span data-line="252"><span class="w">    </span><span class="sd">&quot;&quot;&quot;A query to infer the activity of a promoter in different tissues.</span>
</span><span data-line="253">
</span><span data-line="254"><span class="sd">    Parameters</span>
</span><span data-line="255"><span class="sd">    ----------</span>
</span><span data-line="256"><span class="sd">    promoter_sequence: str</span>
</span><span data-line="257"><span class="sd">        The promoter sequence. Only ATGCN characters are allowed.</span>
</span><span data-line="258"><span class="sd">    orf_sequence: str</span>
</span><span data-line="259"><span class="sd">        The ORF sequence. Only ATGCN characters are allowed.</span>
</span><span data-line="260"><span class="sd">    tissue_of_interest: Dict[str, List[str]]</span>
</span><span data-line="261"><span class="sd">        The tissues of interest, with the tracks representing each tissue, for instance</span>
</span><span data-line="262"><span class="sd">        `{&quot;heart&quot;: [&quot;CNhs10608+&quot;, &quot;CNhs10612+&quot;], &quot;liver&quot;: [&quot;CNhs10608+&quot;, &quot;CNhs10612+&quot;]}`.</span>
</span><span data-line="263"><span class="sd">    query_name: Optional[str] = None</span>
</span><span data-line="264"><span class="sd">        The name of the query. It will appear in the API response and can be used to</span>
</span><span data-line="265"><span class="sd">        handle exceptions.</span>
</span><span data-line="266"><span class="sd">    model: str = &quot;borzoi-human-fold0&quot;</span>
</span><span data-line="267"><span class="sd">        The model to use for the inference (only one default model is supported for now).</span>
</span><span data-line="268">
</span><span data-line="269"><span class="sd">    Returns</span>
</span><span data-line="270"><span class="sd">    -------</span>
</span><span data-line="271"><span class="sd">    PromoterActivityResponse</span>
</span><span data-line="272"><span class="sd">        ``client.send_request(query)`` returns a ``PromoterActivityResponse`` with</span>
</span><span data-line="273"><span class="sd">        attributes ``activity_by_tissue`` (the activity of the promoter in each tissue)</span>
</span><span data-line="274"><span class="sd">        and ``query_name`` (the original query&#39;s name).</span>
</span><span data-line="275"><span class="sd">    &quot;&quot;&quot;</span>
</span><span data-line="276">
</span><span data-line="277">    <span class="n">promoter_sequence</span><span class="p">:</span> <span class="nb">str</span>
</span><span data-line="278">    <span class="n">orf_sequence</span><span class="p">:</span> <span class="nb">str</span>
</span><span data-line="279">    <span class="n">tissue_of_interest</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
</span><span data-line="280">    <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> 
</span><span data-line="281">    <span class="n">model</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;borzoi-human-fold0&quot;</span>
</span><span data-line="282">    <span class="n">query_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span data-line="283">
</span><span data-line="284">    <span class="k">def</span> <span class="nf">to_request_params</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
</span><span data-line="285">        <span class="c1"># TODO: update the web API so the conversion isn&#39;t necessary</span>
</span><span data-line="286">        <span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
</span><span data-line="287">            <span class="s2">&quot;prom&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">promoter_sequence</span><span class="p">,</span>
</span><span data-line="288">            <span class="s2">&quot;orf&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">orf_sequence</span><span class="p">,</span>
</span><span data-line="289">            <span class="s2">&quot;tissue_of_interest&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">tissue_of_interest</span><span class="p">,</span>
</span><span data-line="290">            <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="p">,</span>
</span><span data-line="291">        <span class="p">}</span>
</span><span data-line="292">        <span class="k">return</span> <span class="p">{</span>
</span><span data-line="293">            <span class="s2">&quot;model&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span>
</span><span data-line="294">            <span class="s2">&quot;text&quot;</span><span class="p">:</span> <span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">data</span><span class="p">),</span>
</span><span data-line="295">            <span class="s2">&quot;transforms&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;PROMOTER_ACTIVITY&quot;</span><span class="p">}],</span>
</span><span data-line="296">        <span class="p">}</span>
</span><span data-line="297">
</span><span data-line="298">    <span class="k">def</span> <span class="nf">parse_response</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">results</span><span class="p">):</span>
</span><span data-line="299">        <span class="k">return</span> <span class="n">PromoterActivityResponse</span><span class="p">(</span>
</span><span data-line="300">            <span class="n">query_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">query_name</span><span class="p">,</span> <span class="n">activity_by_tissue</span><span class="o">=</span><span class="n">results</span>
</span><span data-line="301">        <span class="p">)</span>
</span><span data-line="302">
</span><span data-line="303">    <span class="nd">@pydantic</span><span class="o">.</span><span class="n">model_validator</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s2">&quot;after&quot;</span><span class="p">)</span>
</span><span data-line="304">    <span class="k">def</span> <span class="nf">sequences_are_valid_nucleotide_sequences</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">query</span><span class="p">):</span>
</span><span data-line="305"><span class="w">        </span><span class="sd">&quot;&quot;&quot;Raise an error if the sequences contain non-ATGCN characters.&quot;&quot;&quot;</span>
</span><span data-line="306">        <span class="n">query</span><span class="o">.</span><span class="n">promoter_sequence</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">promoter_sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
</span><span data-line="307">        <span class="n">query</span><span class="o">.</span><span class="n">orf_sequence</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">orf_sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
</span><span data-line="308">        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">query</span><span class="o">.</span><span class="n">promoter_sequence</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="s2">&quot;ATGCN&quot;</span><span class="p">)):</span>
</span><span data-line="309">            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
</span><span data-line="310">                <span class="sa">f</span><span class="s2">&quot;Promoter sequence in query &lt;</span><span class="si">{</span><span class="n">query</span><span class="o">.</span><span class="n">query_name</span><span class="si">}</span><span class="s2">&gt; contains &quot;</span>
</span><span data-line="311">                <span class="s2">&quot;non-ATGCN characters.&quot;</span>
</span><span data-line="312">            <span class="p">)</span>
</span><span data-line="313">        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">query</span><span class="o">.</span><span class="n">orf_sequence</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="s2">&quot;ATGCN&quot;</span><span class="p">)):</span>
</span><span data-line="314">            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
</span><span data-line="315">                <span class="sa">f</span><span class="s2">&quot;ORF sequence in query &lt;</span><span class="si">{</span><span class="n">query</span><span class="o">.</span><span class="n">query_name</span><span class="si">}</span><span class="s2">&gt; contains &quot;</span>
</span><span data-line="316">                <span class="s2">&quot;non-ATGCN characters.&quot;</span>
</span><span data-line="317">            <span class="p">)</span>
</span><span data-line="318">        <span class="k">return</span> <span class="n">query</span>
</span><span data-line="319">
</span><span data-line="320">    <span class="nd">@classmethod</span>
</span><span data-line="321">    <span class="k">def</span> <span class="nf">iter_with_promoter_from_fasta</span><span class="p">(</span>
</span><span data-line="322">        <span class="bp">cls</span><span class="p">,</span>
</span><span data-line="323">        <span class="n">fasta_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
</span><span data-line="324">        <span class="n">orf_sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
</span><span data-line="325">        <span class="n">tissue_of_interest</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
</span><span data-line="326">        <span class="n">source</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
</span><span data-line="327">        <span class="n">model</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;borzoi-human-fold0&quot;</span><span class="p">,</span>
</span><span data-line="328">    <span class="p">):</span>
</span><span data-line="329"><span class="w">        </span><span class="sd">&quot;&quot;&quot;Return an iterator of PromoterActivityQuery objects from the promoter</span>
</span><span data-line="330"><span class="sd">        sequences in a fasta file. The iterator has a length attribute that gives the</span>
</span><span data-line="331"><span class="sd">        number of sequences in the fasta file.</span>
</span><span data-line="332">
</span><span data-line="333"><span class="sd">        Parameters</span>
</span><span data-line="334"><span class="sd">        ----------</span>
</span><span data-line="335"><span class="sd">        fasta_path: str</span>
</span><span data-line="336"><span class="sd">            The path to the fasta file containing the promoter sequences.</span>
</span><span data-line="337"><span class="sd">        orf_sequence: str</span>
</span><span data-line="338"><span class="sd">            The ORF sequence.</span>
</span><span data-line="339"><span class="sd">        tissue_of_interest: Dict[str, List[str]]</span>
</span><span data-line="340"><span class="sd">            The tissues of interest, with the tracks representing each tissue, e.g.</span>
</span><span data-line="341"><span class="sd">            `{&quot;heart&quot;: [&quot;CNhs10608+&quot;, &quot;CNhs10612+&quot;], &quot;liver&quot;: [&quot;CNhs10608+&quot;, &quot;CNhs10612+&quot;]}`.</span>
</span><span data-line="342"><span class="sd">        model: str = &quot;borzoi-human-fold0&quot;</span>
</span><span data-line="343"><span class="sd">            The model to use for the inference (only one default model is supported for now).</span>
</span><span data-line="344"><span class="sd">        &quot;&quot;&quot;</span>
</span><span data-line="345">        <span class="n">fasta_iterator</span> <span class="o">=</span> <span class="n">fasta_sequence_iterator</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">)</span>
</span><span data-line="346">        <span class="n">query_iterator</span> <span class="o">=</span> <span class="p">(</span>
</span><span data-line="347">            <span class="bp">cls</span><span class="p">(</span>
</span><span data-line="348">                <span class="n">promoter_sequence</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span>
</span><span data-line="349">                <span class="n">orf_sequence</span><span class="o">=</span><span class="n">orf_sequence</span><span class="p">,</span>
</span><span data-line="350">                <span class="n">tissue_of_interest</span><span class="o">=</span><span class="n">tissue_of_interest</span><span class="p">,</span>
</span><span data-line="351">                <span class="n">source</span><span class="o">=</span><span class="n">source</span><span class="p">,</span>
</span><span data-line="352">                <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span>
</span><span data-line="353">                <span class="n">query_name</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
</span><span data-line="354">            <span class="p">)</span>
</span><span data-line="355">            <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fasta_iterator</span>
</span><span data-line="356">        <span class="p">)</span>
</span><span data-line="357">        <span class="k">return</span> <span class="n">IteratorWithLength</span><span class="p">(</span><span class="n">query_iterator</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">fasta_iterator</span><span class="p">))</span>
</span><span data-line="358">
</span><span data-line="359">    <span class="nd">@classmethod</span>
</span><span data-line="360">    <span class="k">def</span> <span class="nf">list_with_promoter_from_fasta</span><span class="p">(</span>
</span><span data-line="361">        <span class="bp">cls</span><span class="p">,</span>
</span><span data-line="362">        <span class="n">fasta_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
</span><span data-line="363">        <span class="n">orf_sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
</span><span data-line="364">        <span class="n">tissue_of_interest</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
</span><span data-line="365">        <span class="n">source</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
</span><span data-line="366">        <span class="n">model</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;borzoi-human-fold0&quot;</span><span class="p">,</span>
</span><span data-line="367">    <span class="p">):</span>
</span><span data-line="368"><span class="w">        </span><span class="sd">&quot;&quot;&quot;Return a list of PromoterActivityQuery objects from the promoter sequences</span>
</span><span data-line="369"><span class="sd">        in a fasta file.</span>
</span><span data-line="370">
</span><span data-line="371"><span class="sd">        Parameters</span>
</span><span data-line="372"><span class="sd">        ----------</span>
</span><span data-line="373"><span class="sd">        fasta_path: str</span>
</span><span data-line="374"><span class="sd">            The path to the fasta file containing the promoter sequences.</span>
</span><span data-line="375"><span class="sd">        orf_sequence: str</span>
</span><span data-line="376"><span class="sd">            The ORF sequence.</span>
</span><span data-line="377"><span class="sd">        tissue_of_interest: Dict[str, List[str]]</span>
</span><span data-line="378"><span class="sd">            The tissues of interest, with the tracks representing each tissue, e.g.</span>
</span><span data-line="379"><span class="sd">            `{&quot;heart&quot;: [&quot;CNhs10608+&quot;, &quot;CNhs10612+&quot;], &quot;liver&quot;: [&quot;CNhs10608+&quot;, &quot;CNhs10612+&quot;]}`.</span>
</span><span data-line="380"><span class="sd">        model: str = &quot;borzoi-human-fold0&quot;</span>
</span><span data-line="381"><span class="sd">            The model to use for the inference (only one default model is supported for now).</span>
</span><span data-line="382"><span class="sd">        &quot;&quot;&quot;</span>
</span><span data-line="383">        <span class="n">iterator</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">iter_with_promoter_from_fasta</span><span class="p">(</span>
</span><span data-line="384">            <span class="n">fasta_path</span><span class="o">=</span><span class="n">fasta_path</span><span class="p">,</span>
</span><span data-line="385">            <span class="n">orf_sequence</span><span class="o">=</span><span class="n">orf_sequence</span><span class="p">,</span>
</span><span data-line="386">            <span class="n">tissue_of_interest</span><span class="o">=</span><span class="n">tissue_of_interest</span><span class="p">,</span>
</span><span data-line="387">            <span class="n">source</span><span class="o">=</span><span class="n">source</span><span class="p">,</span>
</span><span data-line="388">            <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span>
</span><span data-line="389">        <span class="p">)</span>
</span><span data-line="390">        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">iterator</span><span class="p">)</span>
</span><span data-line="391">
</span><span data-line="392">
</span><span data-line="393"><span class="k">class</span> <span class="nc">DiffusionMaskedResponse</span><span class="p">(</span><span class="n">ResponseBase</span><span class="p">):</span>
</span><span data-line="394"><span class="w">    </span><span class="sd">&quot;&quot;&quot;A response to a DiffusionMaskedQuery, with attributes `sequence` (the predicted</span>
</span><span data-line="395"><span class="sd">    sequence) and `query_name` (the original query&#39;s name).</span>
</span><span data-line="396"><span class="sd">    &quot;&quot;&quot;</span>
</span><span data-line="397">
</span><span data-line="398">    <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span>
</span><span data-line="399">    <span class="n">query_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span data-line="400">
</span><span data-line="401">
</span><span data-line="402"><span class="k">class</span> <span class="nc">DiffusionMaskedQuery</span><span class="p">(</span><span class="n">QueryBase</span><span class="p">):</span>
</span><span data-line="403"><span class="w">    </span><span class="sd">&quot;&quot;&quot;A query to perform masked sampling using a diffusion model.</span>
</span><span data-line="404">
</span><span data-line="405"><span class="sd">    Parameters</span>
</span><span data-line="406"><span class="sd">    ----------</span>
</span><span data-line="407"><span class="sd">    sequence: str</span>
</span><span data-line="408"><span class="sd">        Input sequence for masked sampling. The sequence may contain &quot;&lt;mask&gt;&quot; tokens.</span>
</span><span data-line="409"><span class="sd">    temperature: float, optional (default=0.5)</span>
</span><span data-line="410"><span class="sd">        Sampling temperature, a value between 0 and 1.</span>
</span><span data-line="411"><span class="sd">    decoding_order_strategy: str, optional (default=&quot;entropy&quot;)</span>
</span><span data-line="412"><span class="sd">        Strategy for decoding order, must be either &quot;max_prob&quot; or &quot;entropy&quot;.</span>
</span><span data-line="413"><span class="sd">    unmaskings_per_step: int, optional (default=50)</span>
</span><span data-line="414"><span class="sd">        Number of tokens to unmask per step, an integer between 1 and 1000.</span>
</span><span data-line="415"><span class="sd">    model: str</span>
</span><span data-line="416"><span class="sd">        The model to use for the inference.</span>
</span><span data-line="417"><span class="sd">    query_name: Optional[str] = None</span>
</span><span data-line="418"><span class="sd">        The name of the query. It will appear in the API response and can be used to handle exceptions.</span>
</span><span data-line="419">
</span><span data-line="420"><span class="sd">    Returns</span>
</span><span data-line="421"><span class="sd">    -------</span>
</span><span data-line="422"><span class="sd">    DiffusionMaskedResponse</span>
</span><span data-line="423"><span class="sd">        ``client.send_request(query)`` returns a ``DiffusionMaskedResponse`` with attributes</span>
</span><span data-line="424"><span class="sd">        ``sequence`` (the predicted sequence) and ``query_name`` (the original query&#39;s name).</span>
</span><span data-line="425">
</span><span data-line="426"><span class="sd">    Examples</span>
</span><span data-line="427"><span class="sd">    --------</span>
</span><span data-line="428"><span class="sd">    &gt;&gt;&gt; query = DiffusionMaskedQuery(</span>
</span><span data-line="429"><span class="sd">    ...     sequence=&quot;ATTG&lt;mask&gt;TAC&quot;,</span>
</span><span data-line="430"><span class="sd">    ...     model=&quot;lcdna&quot;,</span>
</span><span data-line="431"><span class="sd">    ...     temperature=0.7,</span>
</span><span data-line="432"><span class="sd">    ...     decoding_order_strategy=&quot;entropy&quot;,</span>
</span><span data-line="433"><span class="sd">    ...     unmaskings_per_step=20,</span>
</span><span data-line="434"><span class="sd">    ... )</span>
</span><span data-line="435"><span class="sd">    &gt;&gt;&gt; client.send_request(query)</span>
</span><span data-line="436"><span class="sd">    DiffusionMaskedResponse(sequence=&quot;ATTGCGTAC&quot;, query_name=None)</span>
</span><span data-line="437"><span class="sd">    &quot;&quot;&quot;</span>
</span><span data-line="438">
</span><span data-line="439">    <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span>
</span><span data-line="440">    <span class="n">temperature</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.5</span>
</span><span data-line="441">    <span class="n">decoding_order_strategy</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;entropy&quot;</span>
</span><span data-line="442">    <span class="n">unmaskings_per_step</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">50</span>
</span><span data-line="443">    <span class="n">model</span><span class="p">:</span> <span class="nb">str</span>
</span><span data-line="444">    <span class="n">query_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span data-line="445">
</span><span data-line="446">    <span class="k">def</span> <span class="nf">to_request_params</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
</span><span data-line="447">        <span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
</span><span data-line="448">            <span class="s2">&quot;sequence&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span>
</span><span data-line="449">            <span class="s2">&quot;temperature&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">temperature</span><span class="p">,</span>
</span><span data-line="450">            <span class="s2">&quot;decoding_order_strategy&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoding_order_strategy</span><span class="p">,</span>
</span><span data-line="451">            <span class="s2">&quot;unmaskings_per_step&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">unmaskings_per_step</span><span class="p">,</span>
</span><span data-line="452">        <span class="p">}</span>
</span><span data-line="453">        <span class="k">return</span> <span class="p">{</span>
</span><span data-line="454">            <span class="s2">&quot;model&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span>
</span><span data-line="455">            <span class="s2">&quot;text&quot;</span><span class="p">:</span> <span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">data</span><span class="p">),</span>
</span><span data-line="456">            <span class="s2">&quot;transforms&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;DIFFUSION_GENERATE&quot;</span><span class="p">}],</span>
</span><span data-line="457">        <span class="p">}</span>
</span><span data-line="458">
</span><span data-line="459">    <span class="k">def</span> <span class="nf">parse_response</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">results</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DiffusionMaskedResponse</span><span class="p">:</span>
</span><span data-line="460">        <span class="k">return</span> <span class="n">DiffusionMaskedResponse</span><span class="p">(</span>
</span><span data-line="461">            <span class="n">sequence</span><span class="o">=</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
</span><span data-line="462">            <span class="n">query_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">query_name</span><span class="p">,</span>
</span><span data-line="463">        <span class="p">)</span>
</span><span data-line="464">
</span><span data-line="465">    <span class="nd">@pydantic</span><span class="o">.</span><span class="n">model_validator</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s2">&quot;after&quot;</span><span class="p">)</span>
</span><span data-line="466">    <span class="k">def</span> <span class="nf">validate_query</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">query</span><span class="p">):</span>
</span><span data-line="467">        <span class="n">sequence</span><span class="p">,</span> <span class="n">model</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="n">query</span><span class="o">.</span><span class="n">model</span>
</span><span data-line="468">        <span class="c1"># Validate sequence and model compatibility</span>
</span><span data-line="469">        <span class="n">_validate_model_and_sequence</span><span class="p">(</span>
</span><span data-line="470">            <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span>
</span><span data-line="471">            <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
</span><span data-line="472">            <span class="n">allow_masks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span data-line="473">        <span class="p">)</span>
</span><span data-line="474">        <span class="c1"># Validate temperature</span>
</span><span data-line="475">        <span class="k">if</span> <span class="ow">not</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">query</span><span class="o">.</span><span class="n">temperature</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">:</span>
</span><span data-line="476">            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;temperature must be between 0 and 1&quot;</span><span class="p">)</span>
</span><span data-line="477">        <span class="c1"># Validate decoding_order_strategy</span>
</span><span data-line="478">        <span class="k">if</span> <span class="n">query</span><span class="o">.</span><span class="n">decoding_order_strategy</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;max_prob&quot;</span><span class="p">,</span> <span class="s2">&quot;entropy&quot;</span><span class="p">]:</span>
</span><span data-line="479">            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
</span><span data-line="480">                <span class="s2">&quot;decoding_order_strategy must be &#39;max_prob&#39; or &#39;entropy&#39;&quot;</span>
</span><span data-line="481">            <span class="p">)</span>
</span><span data-line="482">        <span class="c1"># Validate unmaskings_per_step</span>
</span><span data-line="483">        <span class="k">if</span> <span class="ow">not</span> <span class="mi">1</span> <span class="o">&lt;=</span> <span class="n">query</span><span class="o">.</span><span class="n">unmaskings_per_step</span> <span class="o">&lt;=</span> <span class="mi">1000</span><span class="p">:</span>
</span><span data-line="484">            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;unmaskings_per_step must be between 1 and 1000&quot;</span><span class="p">)</span>
</span><span data-line="485">        <span class="k">return</span> <span class="n">query</span>
</span></pre></div>
        </article><button class="back-to-top" type="button">
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
    <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
  </svg>
  <span>Back to top</span>
</button><div class="navigation flex print:hidden"></div></div>
    </div>
  </main>
</div>
<footer class="sy-foot">
  <div class="sy-foot-inner sy-container mx-auto">
    <div class="sy-foot-reserved md:flex justify-between items-center">
      <div class="sy-foot-copyright"><p>2024, Ginkgo Bioworks</p>
  
  <p>
    Made with
    
    <a href="https://www.sphinx-doc.org/">Sphinx</a> and
    
    <a href="https://shibuya.lepture.com">Shibuya theme</a>.
  </p>
</div>
      <div class="sy-foot-socials"></div>
    </div>
  </div>
</footer>
      <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../../_static/shibuya.js?v=e2e99575"></script></body>
</html>